src/shared/Core/Authentication/MicrosoftAuthentication.cs

```diff
@@ -23,8 +25,66 @@ namespace GitCredentialManager.Authentication
 {
     public interface IMicrosoftAuthentication
     {
-        Task<IMicrosoftAuthenticationResult> GetTokenAsync(string authority, string clientId, Uri redirectUri,
+        /// <summary>
+        /// Acquire an access token for a user principal.
+        /// </summary>
+        /// <param name="authority">Azure authority.</param>
+        /// <param name="clientId">Client ID.</param>
+        /// <param name="redirectUri">Redirect URI for the client.</param>
+        /// <param name="scopes">Set of scopes to request.</param>
+        /// <param name="userName">Optional user name for an existing account.</param>
+        /// <param name="msaPt">Use MSA-Passthrough behavior when authenticating.</param>
+        /// <returns>Authentication result.</returns>
+        Task<IMicrosoftAuthenticationResult> GetTokenForUserAsync(string authority, string clientId, Uri redirectUri,
             string[] scopes, string userName, bool msaPt = false);
+
+        /// <summary>
+        /// Acquire an access token for the given service principal with the specified scopes.
+        /// </summary>
+        /// <param name="sp">Service principal identity.</param>
+        /// <param name="scopes">Scopes to request.</param>
+        /// <returns>Authentication result.</returns>
+        Task<IMicrosoftAuthenticationResult> GetTokenForServicePrincipalAsync(ServicePrincipalIdentity sp, string[] scopes);
+
+        /// <summary>
+        /// Acquire a token using the managed identity in the current environment.
+        /// </summary>
+        /// <param name="managedIdentity">
+        /// Client ID of the user-assigned managed identity.
```

Comment:
Excellent abstraction :)

Reply:
FYI - In MSAL we went with a more verbose API because you can specify the MI in different ways: 

- by client id
- by resource id
- by object id
- by SP id

There are scenarios where you want to use resource id instead of client id. But anyway, you can address the scenarios later.

Reply:
From my understanding, a MI is really just an SP under the hood. Is this correct? It's how it appears when you go digging in the AAD portal.

Also it's my understanding that:

- Client ID = ID of the "application"
- SP ID = ID of the materialised SP in a particular tenant

The application (or MI) owns the credentials (secret or cert).

What is the resource ID? Could this be, for example, the object ID of the Virtual Machine resource that has a system-assigned MI? I can certainly see how this may be useful (since you may not know what the system MI OID is for a resource, like a VM).

How does that work for user-assigned MIs?

The MSAL API only shows me two options:

- `WithUserAssignedClientId`
- `WithUserAssignedResourceId`